t/proxyunits.t

```diff
@@ -411,16 +432,55 @@ check_version($ps);
     print $be "TOUCHED\r\n";
 
     is(scalar <$ps>, "TOUCHED\r\n", "got TOUCHED instread of STORED");
+}
+
+check_version($ps);
 
-    # TODO: meta quiet cases
-    # - q should be turned into a space on the backend
-    # - errors should still pass through to client
+{
+    announce("Test quiet flag", __LINE__);
+
+    # make sure q is replaced by a space, and HD is not returned to client
+    my $be = $mbe[0];
+    my $cmd = "ms /b/a 2 q\r\nhi\r\n";
+    print $ps $cmd;
+    is(scalar <$be>, "ms /b/a 2  \r\n", "set received with q replaced by a space");
+    is(scalar <$be>, "hi\r\n", "set payload received");
+
+    print $be "HD\r\n";
+
+    check_version($ps);
```

Comment:
I've typically only been running check_version inbetween each block. though I guess it isn't terrible to do it more often?

Reply:
check_version($ps) is great, one caveat is it does not check backends, I am thinking to evolve it to check both.

Reply:
It doesn't always make sense to check the backends; and the test will still fail if there's junk in there or the backend is dead. I wouldn't want to add too much re-checking between sub tests as it makes the tests too slow.

Reply:
https://github.com/memcached/memcached/pull/1030

It seems to be a good idea to me to clean up and restore the environment between groups of test scenarios? So folks can have a peace of mind to add more tests without worrying about pollutions from previous ones.

Reply:
I'd push the opposite direction actually: you should usually clean up a test deliberately if you know it's being left in a bad state. Otherwise we can paper over real bugs, as I've found while writing these scenarios.

ie; I had a regression where the final test in `proxyconfig.t` should have been failing, but it didn't until I changed my environment. Had I been cleaning up the environment after batches of tests it would've never failed.

Basically I try to write groups of tests such that they roll through clean states. Real user environments don't clean up everything as they work.

Reply:
> you should usually clean up a test deliberately if you know it's being left in a bad state. Otherwise we can paper over real bugs, as I've found while writing these scenarios.

I agree with that, and that's exactly why a bug is found thanks to this change, and it serves as a forcing function to make sure everyone cleanups properly otherwise your new tests will fail.

> Basically I try to write groups of tests such that they roll through clean states. Real user environments don't clean up everything as they work.

Can this be an intermediate state before we get there, or you prefer to have this right now?

Reply:
What was the bug that got fixed in this example? You said "a bug" but it wasn't obvious what changed or why.

Reply:
https://github.com/memcached/memcached/pull/1030/files#diff-438bf7a9c2561cba34fed3bd2b39404deb1c3fa9891d4f65715d85c64403f6e6R676

The 3rd backend was accidentally killed due to not sending response back.  It took me a while to debug, and have check_sanity() make sure this got caught.

Reply:
Ok, I see. thanks!

To summarize my thoughts:

- checking between broader sections is totally fine. I was already doing that part way anyway.
- Checking between every sub-test is not always the right thing to do; you can throw it in more while developing and should use it liberally when debugging. However I've found real code bugs (not test bugs) because feature A used after feature B didn't work as expected. check_sanity() is fine since it's not re-establishing the backend but much more than that isn't helpful.

If we tear everything down all the time we miss a lot of potential bugs, or have to write far more total tests for coverage.

Reply:
Thanks for the summary. We are on the page! `check_sanity` is meant to iterate on `check_versions` that checks between boarder sections.

Reply:
Sounds good. You had a section of code where you were running it like 8 times which is why I brought it up.

With proxyconfig.t I struggled with this same issue; to get clean backends I had to issue a proxy reload in a way where the backend labels change so it internally creates new objects and re-connects. It's not abstracted yet.